### PR TITLE
Add charset support for RFC 7617 challenges

### LINKED ISSUE #1468: Charset support for Basic Auth
http://datatracker.ietf.org/doc/draft-ietf-httpauth-basicauth-update/?include_text=1

### PR DESCRIPTION
**Pull request by erikghonyan**
Fixes #1468

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/Challenge.java [hunk 1]
```diff
@@ -17,14 +17,16 @@
...
 
-/** An RFC 2617 challenge. */
+/** An RFC 7617 challenge. */
 public final class Challenge {
   private final String scheme;
   private final String realm;
+  private final String charset;
 
-  public Challenge(String scheme, String realm) {
+  public Challenge(String scheme, String realm, String charset) {
```
Comment (erikghonyan) [inline okhttp/src/main/java/okhttp3/Challenge.java]: Do we need a constructor with just scheme and realm parameters?

**Comment by erikghonyan** [comment]
Should I add a test for this or should I send a separate pull request for that?

File: okhttp/src/main/java/okhttp3/Challenge.java (hunk 1 above)
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/Challenge.java]: We definitely need to keep the original constructor for compatibility. I was thinking something like this:

```
Challenge c = new Challenge("basic", "WallyWorld");
c = c.withCharset(StandardCharsets.UTF_8);
```

Basically keep the public constructor as-is and add a public factory method that adds a charset to an existing challenge.

File: okhttp/src/main/java/okhttp3/internal/http/HttpHeaders.java [hunk 2]
```diff
@@ -158,17 +158,27 @@ public static Headers varyHeaders(Headers requestHeaders, Headers responseHeader
...
         if (header.regionMatches(true, matcher.start(1), "realm", 0, 5)) {
-          String scheme = header.substring(0, index);
-          String realm = matcher.group(3);
-          if (realm != null) {
-            challenges.add(new Challenge(scheme, realm));
-            break;
-          }
+          realm = matcher.group(3);
+        } else if (header.regionMatches(true, matcher.start(1), "charset", 0, 7)) {
+          charset = matcher.group(3);
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http/HttpHeaders.java]: dig

Comment (erikghonyan) [inline okhttp/src/main/java/okhttp3/internal/http/HttpHeaders.java]: Can you provide more details please?

**Comment by swankjesse** [comment]
Merged manually. Thanks for your patience, @erikghonyan !

**Comment by swankjesse** [comment]
https://github.com/square/okhttp/commit/b3ecb0a6c9bf3a213377b889f2b3077a940363e2

**Pull request closed without merging**